[TSPLIBImporter] Consider multi-space delimiters (#1060)

HannesWell opened the pull request:
This PR provides a patch to fix #1060 and adjusts the tests to cover the cases of the mentioned issue.

Additionally I simplified the default implementation of `GraphImporter.importGraph(Graph<V, E> g, File file)` by using a `FileReader`.

<!-- describe the changes you have made here: what, why, …
     Link issues by using the following pattern: [#333](https://github.com/jgrapht/jgrapht/issues/333).
     The title of the PR must not reference an issue, because GitHub does not support autolinking there. -->

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [ ] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

HannesWell commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -619,9 +618,11 @@ public void importGraph(Graph<V, E> graph, Reader in)
         return nodes;
     }
 
+    private static final Pattern WHITE_SPACE = Pattern.compile("[ \t]+");
```
The pattern "[ \t]+" allows an arbitrary number (greater zero) of spaces and tabs to be used as delimiter.

Instead of " \t" the predefined character class "\\s+" could be used to allow arbitrary white-spaces.
From the java-doc of the `java.util.regex.Pattern` class:
```
Predefined character classes
\s  |  A whitespace character: [ \t\n\x0B\f\r]
```

I think it is very unlikely that arbitrary white-spaces are used (especially since \n and \r are used as line-separator).
What do you think?

jkinable replied:
Looks fine to me. Unlikely that anything else than tabs or spaces will be used.

jkinable commented:
Just to make sure, I would download [all TSPLib instances](http://comopt.ifi.uni-heidelberg.de/software/TSPLIB95/tsp/ALL_tsp.tar.gz), and simply run them through the importer in a for loop to make sure that there are no other problematic instances.

jkinable commented:
Thanks for the quick fix btw :)

HannesWell commented:
You are very welcome. It's nice to see that the importer is also used by somebody else (even I just know it due to a bug).

I just wrote the following little code to check all the instances in the given link.
Requires Apache commons-compress: https://mvnrepository.com/artifact/org.apache.commons/commons-compress

```
package org.jgrapht.nio;

import org.apache.commons.compress.archivers.*;
import org.apache.commons.compress.archivers.tar.*;
import org.apache.commons.compress.compressors.gzip.*;
import org.jgrapht.*;
import org.jgrapht.graph.*;
import org.jgrapht.nio.tsplib.*;
import org.jgrapht.nio.tsplib.TSPLIBImporter.*;
import org.jgrapht.util.*;

import java.io.*;
import java.net.*;
import java.util.regex.*;

public class TestAll
{
    private static final int MAX_INSTANCE_SIZE = 2000;

    public static void main(String[] args)
        throws IOException
    {
        String url = "http://comopt.ifi.uni-heidelberg.de/software/TSPLIB95/tsp/ALL_tsp.tar.gz";
        InputStream in = new BufferedInputStream(new URL(url).openStream());
        try (
            ArchiveInputStream tar = new TarArchiveInputStream(new GzipCompressorInputStream(in)))
        {
            for (ArchiveEntry entry; (entry = tar.getNextEntry()) != null;) {
                testTSPLIBInstance(new GzipCompressorInputStream(tar), entry.getName());
            }
        }
    }

    private static final Pattern TSP_FILE_PATTERN = Pattern.compile("[a-zA-Z]+(\\d+)\\.tsp\\.gz");
    private static final int SIZE_GROUP = 1;

    private static void testTSPLIBInstance(InputStream inStream, String name)
    {
        Matcher matcher = TSP_FILE_PATTERN.matcher(name);
        if (!matcher.matches()) {
            return;
        }
        int expectedSize = Integer.parseInt(matcher.group(SIZE_GROUP));
        if (expectedSize > MAX_INSTANCE_SIZE) {
            System.out.println("Skip test " + name + " (too large)");
            return;
        }

        System.out.println("Test " + name);

        Graph<Integer, DefaultWeightedEdge> graph = new SimpleDirectedWeightedGraph<>(
            SupplierUtil.createIntegerSupplier(), SupplierUtil.createDefaultWeightedEdgeSupplier());

        try {
            TSPLIBImporter<Integer, DefaultWeightedEdge> importer = new TSPLIBImporter<>();
            importer.importGraph(graph, inStream);

            assertSize(name + " - vertexSet", expectedSize, graph.vertexSet().size());
            Metadata<Integer, DefaultWeightedEdge> metadata = importer.getMetadata();
            assertSize(
                name + " - metadata", expectedSize, metadata.getVertexToNodeMapping().size());
        } catch (Exception e) {
            System.out.println("Failed : " + e.getMessage());
        }
    }

    public static void assertSize(String name, int expectedSize, int actualSize)
    {
        if (actualSize != expectedSize) {
            throw new IllegalStateException(
                "Instance " + name + " expected " + expectedSize + " and read " + actualSize);
        }
    }
}
```

The following output is written:
```
Test a280.tsp.gz
Test ali535.tsp.gz
Test att48.tsp.gz
Test att532.tsp.gz
Test bayg29.tsp.gz
Failed : Instance bayg29.tsp.gz - vertexSet expected 29 and read 0
Test bays29.tsp.gz
Failed : Instance bays29.tsp.gz - vertexSet expected 29 and read 0
Test berlin52.tsp.gz
Test bier127.tsp.gz
Test brazil58.tsp.gz
Failed : Instance brazil58.tsp.gz - vertexSet expected 58 and read 0
Skip test brd14051.tsp.gz (too large)
Test brg180.tsp.gz
Failed : Instance brg180.tsp.gz - vertexSet expected 180 and read 0
Test burma14.tsp.gz
Test ch130.tsp.gz
Test ch150.tsp.gz
Test d1291.tsp.gz
Skip test d15112.tsp.gz (too large)
Test d1655.tsp.gz
Skip test d18512.tsp.gz (too large)
Test d198.tsp.gz
Skip test d2103.tsp.gz (too large)
Test d493.tsp.gz
Test d657.tsp.gz
Test dantzig42.tsp.gz
Failed : Instance dantzig42.tsp.gz - vertexSet expected 42 and read 0
Test dsj1000.tsp.gz
Test eil101.tsp.gz
Test eil51.tsp.gz
Test eil76.tsp.gz
Test fl1400.tsp.gz
Test fl1577.tsp.gz
Skip test fl3795.tsp.gz (too large)
Test fl417.tsp.gz
Skip test fnl4461.tsp.gz (too large)
Test fri26.tsp.gz
Failed : Instance fri26.tsp.gz - vertexSet expected 26 and read 0
Test gil262.tsp.gz
Test gr120.tsp.gz
Failed : Instance gr120.tsp.gz - vertexSet expected 120 and read 0
Test gr137.tsp.gz
Test gr17.tsp.gz
Failed : Instance gr17.tsp.gz - vertexSet expected 17 and read 0
Test gr202.tsp.gz
Test gr21.tsp.gz
Failed : Instance gr21.tsp.gz - vertexSet expected 21 and read 0
Test gr229.tsp.gz
Test gr24.tsp.gz
Failed : Instance gr24.tsp.gz - vertexSet expected 24 and read 0
Test gr431.tsp.gz
Test gr48.tsp.gz
Failed : Instance gr48.tsp.gz - vertexSet expected 48 and read 0
Test gr666.tsp.gz
Test gr96.tsp.gz
Test hk48.tsp.gz
Failed : Instance hk48.tsp.gz - vertexSet expected 48 and read 0
Test kroA100.tsp.gz
Test kroA150.tsp.gz
Test kroA200.tsp.gz
Test kroB100.tsp.gz
Test kroB150.tsp.gz
Test kroB200.tsp.gz
Test kroC100.tsp.gz
Test kroD100.tsp.gz
Test kroE100.tsp.gz
Test lin105.tsp.gz
Test lin318.tsp.gz
Test linhp318.tsp.gz
Test nrw1379.tsp.gz
Test p654.tsp.gz
Test pa561.tsp.gz
Failed : Instance pa561.tsp.gz - vertexSet expected 561 and read 0
Test pcb1173.tsp.gz
Skip test pcb3038.tsp.gz (too large)
Test pcb442.tsp.gz
Skip test pla33810.tsp.gz (too large)
Skip test pla7397.tsp.gz (too large)
Skip test pla85900.tsp.gz (too large)
Test pr1002.tsp.gz
Test pr107.tsp.gz
Test pr124.tsp.gz
Test pr136.tsp.gz
Test pr144.tsp.gz
Test pr152.tsp.gz
Test pr226.tsp.gz
Skip test pr2392.tsp.gz (too large)
Test pr264.tsp.gz
Test pr299.tsp.gz
Test pr439.tsp.gz
Test pr76.tsp.gz
Test rat195.tsp.gz
Test rat575.tsp.gz
Test rat783.tsp.gz
Test rat99.tsp.gz
Test rd100.tsp.gz
Test rd400.tsp.gz
Skip test rl11849.tsp.gz (too large)
Test rl1304.tsp.gz
Test rl1323.tsp.gz
Test rl1889.tsp.gz
Skip test rl5915.tsp.gz (too large)
Skip test rl5934.tsp.gz (too large)
Test si1032.tsp.gz
Failed : Failed to import graph from TSPLIB-file: Invalid TYPE value <TSP (M.~Hofmeister)>
Test si175.tsp.gz
Failed : Failed to import graph from TSPLIB-file: Invalid TYPE value <TSP (M.~Hofmeister)>
Test si535.tsp.gz
Failed : Failed to import graph from TSPLIB-file: Invalid TYPE value <TSP (M.~Hofmeister)>
Test st70.tsp.gz
Test swiss42.tsp.gz
Failed : Instance swiss42.tsp.gz - vertexSet expected 42 and read 0
Test ts225.tsp.gz
Test tsp225.tsp.gz
Test u1060.tsp.gz
Test u1432.tsp.gz
Test u159.tsp.gz
Test u1817.tsp.gz
Skip test u2152.tsp.gz (too large)
Skip test u2319.tsp.gz (too large)
Test u574.tsp.gz
Test u724.tsp.gz
Test ulysses16.tsp.gz
Test ulysses22.tsp.gz
Skip test usa13509.tsp.gz (too large)
Test vm1084.tsp.gz
Test vm1748.tsp.gz
```

I checked some (not all) of those instances that fail because no nodes were read and they all have an _EDGE_WEIGHT_SECTION_ instead of a _NODE_COORD_SECTION_. But an EDGE_WEIGHT_SECTION is not yet supported.

The failure because of an invalid TYPE values is also correct if we strictly follow the spec. But being more forgiving could be implemented quite-quickly (simply discard everything after the first white-space). Nevertheless those instance would also fail because they use a EDGE_WEIGHT_SECTION.

jkinable commented:
Yes, please fix those. It would be really odd to have a TSPLib importer.... that fails to read various of the standard TSPLib instances...

Btw, what's the challenge with the EDGE_WEIGHT_SECTION? Is there anything specific that blocks us from implementing this?
Here's another Java based TSPLib parser: https://github.com/coin-or/jorlib/tree/master/jorlib-core/src/main/java/org/jorlib/io/tspLibReader
Admittedly, this parser isn't particularly great, but it might serve as a reference.

HannesWell commented:
I changed the importer so it ignores everything after the first white-space in the value part while reading a specification element that requires a certain value literal. For example while reading the line `TYPE : TSP (Some comment)`, everything after the TSP literal is ignored. To be precise the trimmed value is now split at white-spaces of arbitrary length and only the first element of the resulting array is used as value.

HannesWell commented:
I think there should not be a real challenge in implementing this. It was simply not needed, therefore I only implemented the required parts.

But you are right. It would be better if the full standard is supported. IIRC there were some parts that weren't feasible with JGraphT, but I have to look into the details first. 
Since this would be a bigger change I suggest to open another Issue/PR for that.

jkinable commented:
@HannesWell PR looks good to me. I'll merge as-is. Could you re-run the test on all the instances you ran a few days ago and post the result? That way we can keep track of the current state.
In general it seems that the TSPLIB importer could benefit from some additional testing, to ensure that we don't run into more issues such as the one in this PR or as in issue #1064.

Thanks again for your effort.

HannesWell commented:
Thanks.

I will run the test on all instances tomorrow and post the results to #1067 and will see what needs to be fixed.

Regarding the test coverage. What I see is mainly that creators of tsplib files seem to take a lot of freedom and do not follow the specification exactly. For example those that failed with the message: `Invalid TYPE value <TSP (M.~Hofmeister)>`
The specification only allows certain values for an entry with key `TYPE`. One of them is `TSP`, but subsequent comments are not intended, at least I did not find anything about that in the spec. For comments there is actually a separate entry with keyword `comment`.
Nevertheless the tests can be extended to also cover these slightly off-spec files like I already did in this PR.
